### PR TITLE
Allow sorting the loadsave savegame list

### PR DESCRIPTION
**Pull request by Noordfrees**
image

### CODE REVIEW

File: src/loadsave_gui.cpp [hunk 1]
```diff
@@ -15,6 +18,34 @@
 #include "rev.h"
 #include "sprite_store.h"
 
+/** Sorting order for a list of savegames. */
+using SortBy = std::function<bool(const PreloadData&, const PreloadData&)>;
```
Comment (LordAro) [inline src/loadsave_gui.cpp]: Generally, std::function should be avoided due to its high overheads and general slowness https://wolchok.org/posts/cxx-trap-2-std-function/

File: src/loadsave_gui.cpp [hunk 2]
```diff
@@ -15,6 +18,34 @@
...
+
+/** Sort by timestamp. */
+static const SortBy SORT_BY_TIMESTAMP = [](const PreloadData &a, const PreloadData &b) {
```
Comment (LordAro) [inline src/loadsave_gui.cpp]: Top-level lambdas are a bit... weird. Why not just use actual functions?

File: src/loadsave_gui.cpp [hunk 3]
```diff
@@ -32,12 +63,14 @@ class LoadSaveGui : public GuiWindow {
...
+	void Sort(const SortBy &sort);
 
 private:
 	std::string FinalFilename() const;
 
-	const Type type;                  ///< Type of this window.
-	std::set<PreloadData> all_files;  ///< All files in the working directory.
+	const Type type;                   ///< Type of this window.
+	std::list<PreloadData> all_files;  ///< All files in the working directory.
+	const SortBy *current_sort;        ///< Current sorting order.
```
Comment (LordAro) [inline src/loadsave_gui.cpp]: Avoiding std::function does mean you have to use void* to store the function pointer which isn't so nice for the typesystem, but suitable amount of templating (on Sort) should get you pretty close

File: src/loadsave_gui.cpp [hunk 4]
```diff
@@ -32,12 +63,14 @@ class LoadSaveGui : public GuiWindow {
...
 	bool OnKeyEvent(WmKeyCode key_code, WmKeyMod mod, const std::string &symbol) override;
+	void Sort(const SortBy &sort);
 
 private:
 	std::string FinalFilename() const;
 
-	const Type type;                  ///< Type of this window.
-	std::set<PreloadData> all_files;  ///< All files in the working directory.
+	const Type type;                   ///< Type of this window.
+	std::list<PreloadData> all_files;  ///< All files in the working directory.
```
Comment (LordAro) [inline src/loadsave_gui.cpp]: std::list ? [Linked list bad](https://stackoverflow.com/a/34171273/995325), just use std::vector

Comment (LordAro) [inline src/loadsave_gui.cpp]: std::list ? [Linked list bad](https://stackoverflow.com/a/34171273/995325), just use std::vector

**Comment by Alberth289346** [comment]
Done

**Pull request merged**